File: src/PicLens.Infrastructure/Services/FolderScanner.cs
using PicLens.Core.Services;
using PicLens.Core.Domain;
using PicLens.Core.Models;
using SkiaSharp;

namespace PicLens.Infrastructure.Services;

public sealed class FolderScanner : IFolderScanner
{
    private static readonly EnumerationOptions SafeEnumerationOptions = new()
    {
        AttributesToSkip = 0,
        IgnoreInaccessible = true
    };

    public Task<IReadOnlyList<ListItem>> ScanAsync(ListQuery query, CancellationToken cancellationToken = default)
    {
        return Task.Run(() => Scan(query, cancellationToken), cancellationToken);
    }

    public Task<IReadOnlyList<FolderListItem>> ScanChildFoldersAsync(
        string folderPath,
        CancellationToken cancellationToken = default)
    {
        return Task.Run(() => ScanChildFolders(folderPath, cancellationToken), cancellationToken);
    }

    private static IReadOnlyList<ListItem> Scan(ListQuery query, CancellationToken cancellationToken)
    {
        if (!Directory.Exists(query.FolderPath))
        {
            throw new DirectoryNotFoundException(query.FolderPath);
        }

        var items = query.IncludeSubfolders
            ? EnumerateRecursiveImages(query.FolderPath, cancellationToken).Cast<ListItem>().ToList()
            : EnumerateDirectItems(query.FolderPath, cancellationToken).ToList();

        var sorted = ListItemSorter.Sort(items, query.Sort, keepFoldersFirst: !query.IncludeSubfolders);
        return sorted;
    }

    private static IReadOnlyList<FolderListItem> ScanChildFolders(
        string folderPath,
        CancellationToken cancellationToken)
    {
        if (!Directory.Exists(folderPath))
        {
            throw new DirectoryNotFoundException(folderPath);
        }

        var folders = EnumerateDirectFolders(folderPath, cancellationToken).Cast<ListItem>().ToList();
        var sorted = ListItemSorter.Sort(
                folders,
                new SortState(SortKey.Name, SortDirection.Asc),
                keepFoldersFirst: false)
            .OfType<FolderListItem>()
            .ToList();
        return sorted;
    }

    private static IEnumerable<ListItem> EnumerateDirectItems(string folderPath, CancellationToken cancellationToken)
    {
        foreach (var folder in EnumerateDirectFolders(folderPath, cancellationToken))
        {
            yield return folder;
        }

        foreach (var file in SafeEnumerateFiles(folderPath))
        {
            cancellationToken.ThrowIfCancellationRequested();
            var image = CreateImageItem(file);
            if (image is not null)
            {
                yield return image;
            }
        }
    }

    private static IEnumerable<FolderListItem> EnumerateDirectFolders(string folderPath, CancellationToken cancellationToken)
    {
        foreach (var directory in SafeEnumerateDirectories(folderPath))
        {
            cancellationToken.ThrowIfCancellationRequested();
            yield return new FolderListItem(
                Path: directory,
                Name: Path.GetFileName(directory),
                ModifiedAtMs: ToUnixMs(Directory.GetLastWriteTimeUtc(directory)));
        }
    }

    private static IEnumerable<ImageListItem> EnumerateRecursiveImages(string root, CancellationToken cancellationToken)
    {
        var pending = new Stack<string>();
        var visited = new HashSet<string>(PathKeyComparer);
        pending.Push(root);

        while (pending.Count > 0)
        {
            cancellationToken.ThrowIfCancellationRequested();
            var current = pending.Pop();
            var canonical = CanonicalDirectoryKey(current);
            if (canonical is null || !visited.Add(canonical))
            {
                continue;
            }

            foreach (var directory in SafeEnumerateDirectories(current))
            {
                pending.Push(directory);
            }

            foreach (var file in SafeEnumerateFiles(current))
            {
                cancellationToken.ThrowIfCancellationRequested();
                var image = CreateImageItem(file);
                if (image is not null)
                {
                    yield return image;
                }
            }
        }
    }

    private static ImageListItem? CreateImageItem(string file)
    {
        var extension = ImageFormatRules.GetSupportedImageExtension(file);
        if (extension is null)
        {
            return null;
        }

        try
        {
            var info = new FileInfo(file);
            var isAnimated = RequiresAnimationDetection(extension)
                && IsKnownAnimatedImage(file);

            return new ImageListItem(
                Path: file,
                Name: info.Name,
                Extension: extension,
                ModifiedAtMs: ToUnixMs(info.LastWriteTimeUtc),
                SizeBytes: info.Length,
                IsAnimated: isAnimated);
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            return null;
        }
    }

    private static bool RequiresAnimationDetection(string extension) =>
        extension.Equals("gif", StringComparison.OrdinalIgnoreCase)
        || extension.Equals("webp", StringComparison.OrdinalIgnoreCase);

    private static bool IsKnownAnimatedImage(string path)
    {
        try
        {
            using var stream = OpenProbeStream(path);
            using var codec = SKCodec.Create(stream);
            return codec?.FrameCount > 1;
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            return false;
        }
    }

    private static FileStream OpenProbeStream(string path) =>
        new(path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite | FileShare.Delete);

    private static IEnumerable<string> SafeEnumerateDirectories(string folderPath) =>
        Directory.EnumerateDirectories(folderPath, "*", SafeEnumerationOptions);

    private static IEnumerable<string> SafeEnumerateFiles(string folderPath) =>
        Directory.EnumerateFiles(folderPath, "*", SafeEnumerationOptions);

    private static long ToUnixMs(DateTime value) =>
        new DateTimeOffset(DateTime.SpecifyKind(value, DateTimeKind.Utc)).ToUnixTimeMilliseconds();

    private static StringComparer PathKeyComparer => PathRules.PathComparer;

    private static string? CanonicalDirectoryKey(string folderPath)
    {
        try
        {
            var directory = new DirectoryInfo(folderPath);
            var resolved = directory.ResolveLinkTarget(returnFinalTarget: true);
            return Path.GetFullPath(resolved?.FullName ?? directory.FullName);
        }
        catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
        {
            return null;
        }
    }
}

File: src/infrastructure/src/folder_scanner.cpp
#include <piclens/infrastructure/folder_scanner.h>

#include <piclens/core/image_format_rules.h>
#include <piclens/core/list_item_sorter.h>
#include <piclens/core/path_rules.h>
#include <piclens/infrastructure/image_animation_probe.h>

#include <QDir>
#include <QFileInfo>
#include <QSet>

namespace piclens::infrastructure {
namespace {

QDir::Filters directoryFilters()
{
    return QDir::Dirs | QDir::Hidden | QDir::System | QDir::NoDotAndDotDot;
}

QDir::Filters fileFilters()
{
    return QDir::Files | QDir::Hidden | QDir::System;
}

QString canonicalDirectoryKey(const QString &path)
{
    QString canonical = QFileInfo(path).canonicalFilePath();
    if (canonical.isEmpty()) {
        return {};
    }
    canonical = QDir::cleanPath(canonical);
    return core::path_rules::pathCaseSensitivity() == Qt::CaseInsensitive
        ? canonical.toCaseFolded()
        : canonical;
}

} // namespace

DirectoryNotFoundError::DirectoryNotFoundError(const QString &path)
    : std::runtime_error(QStringLiteral("Directory not found: %1").arg(path).toStdString())
{
}

ScanCanceledError::ScanCanceledError()
    : std::runtime_error("Folder scan canceled.")
{
}

QVector<core::ListItem> FolderScanner::scan(
    const core::ListQuery &query,
    std::stop_token stopToken) const
{
    throwIfCanceled(stopToken);
    const QFileInfo rootInfo(query.folderPath);
    if (!rootInfo.exists() || !rootInfo.isDir()) {
        throw DirectoryNotFoundError(query.folderPath);
    }

    QVector<core::ListItem> items;
    if (!query.includeSubfolders) {
        const auto folders = directFolders(query.folderPath, stopToken);
        items.reserve(folders.size());
        for (const auto &folder : folders) {
            items.append(folder);
        }

        for (const QFileInfo &file : directFiles(query.folderPath)) {
            throwIfCanceled(stopToken);
            if (auto image = createImageItem(file); image.has_value()) {
                items.append(std::move(*image));
            }
        }
    } else {
        QVector<QString> pending{rootInfo.absoluteFilePath()};
        QSet<QString> visited;

        while (!pending.isEmpty()) {
            throwIfCanceled(stopToken);
            const QString current = pending.takeLast();
            const QString canonical = canonicalDirectoryKey(current);
            if (canonical.isEmpty() || visited.contains(canonical)) {
                continue;
            }
            visited.insert(canonical);

            const auto folders = directFolders(current, stopToken);
            for (const auto &folder : folders) {
                pending.append(folder.path);
            }
            for (const QFileInfo &file : directFiles(current)) {
                throwIfCanceled(stopToken);
                if (auto image = createImageItem(file); image.has_value()) {
                    items.append(std::move(*image));
                }
            }
        }
    }

    return core::list_item_sorter::sort(
        items,
        query.sort,
        !query.includeSubfolders);
}

QVector<core::FolderListItem> FolderScanner::scanChildFolders(
    const QString &folderPath,
    std::stop_token stopToken) const
{
    throwIfCanceled(stopToken);
    const QFileInfo rootInfo(folderPath);
    if (!rootInfo.exists() || !rootInfo.isDir()) {
        throw DirectoryNotFoundError(folderPath);
    }

    QVector<core::ListItem> items;
    const auto folders = directFolders(folderPath, stopToken);
    items.reserve(folders.size());
    for (const auto &folder : folders) {
        items.append(folder);
    }

    const auto sorted = core::list_item_sorter::sort(
        items,
        {.key = core::SortKey::Name, .direction = core::SortDirection::Asc},
        false);
    QVector<core::FolderListItem> result;
    result.reserve(sorted.size());
    for (const auto &item : sorted) {
        result.append(std::get<core::FolderListItem>(item));
    }
    return result;
}

void FolderScanner::throwIfCanceled(std::stop_token stopToken)
{
    if (stopToken.stop_requested()) {
        throw ScanCanceledError();
    }
}

QVector<core::FolderListItem> FolderScanner::directFolders(
    const QString &folderPath,
    std::stop_token stopToken)
{
    const QFileInfoList entries = QDir(folderPath).entryInfoList(directoryFilters(), QDir::NoSort);
    QVector<core::FolderListItem> folders;
    folders.reserve(entries.size());
    for (const QFileInfo &entry : entries) {
        throwIfCanceled(stopToken);
        folders.append({
            .path = entry.absoluteFilePath(),
            .name = entry.fileName(),
            .modifiedAtMs = entry.lastModified().toMSecsSinceEpoch(),
        });
    }
    return folders;
}

QVector<QFileInfo> FolderScanner::directFiles(const QString &folderPath)
{
    return QDir(folderPath).entryInfoList(fileFilters(), QDir::NoSort);
}

std::optional<core::ImageListItem> FolderScanner::createImageItem(const QFileInfo &file)
{
    const auto extension = core::image_format_rules::supportedImageExtension(file.absoluteFilePath());
    if (!extension.has_value()) {
        return std::nullopt;
    }

    const bool requiresAnimationProbe = extension->compare(QStringLiteral("gif"), Qt::CaseInsensitive) == 0
        || extension->compare(QStringLiteral("webp"), Qt::CaseInsensitive) == 0;
    return core::ImageListItem{
        .path = file.absoluteFilePath(),
        .name = file.fileName(),
        .extension = *extension,
        .modifiedAtMs = file.lastModified().toMSecsSinceEpoch(),
        .sizeBytes = file.size(),
        .isAnimated = requiresAnimationProbe
            && image_animation_probe::isAnimated(file.absoluteFilePath()),
    };
}

} // namespace piclens::infrastructure
